Readability.js

```diff
@@ -844,18 +844,50 @@ Readability.prototype = {
 
         this._initializeNode(topCandidate);
       } else if (topCandidate) {
+        parentOfTopCandidate = topCandidate.parentNode;
+
+        // Find a better top candidate node if it contains (at least three) nodes which belong to `topCandidates` array
+        // and whose scores are quite closed with current `topCandidate` node.
+        var alternativeCandidateAncestors = [];
+        for (var i = 1; i < topCandidates.length; i++) {
+          if (topCandidates[i].readability.contentScore / topCandidate.readability.contentScore >= 0.75) {
+            alternativeCandidateAncestors.push(this._getNodeAncestors(topCandidates[i]));
+          }
+        }
+        var MINIMUM_TOPCANDIDATES = 3;
+        if (alternativeCandidateAncestors.length >= MINIMUM_TOPCANDIDATES) {
+          while (parentOfTopCandidate && parentOfTopCandidate.tagName !== "BODY") {
+            if (!parentOfTopCandidate.readability) {
+              parentOfTopCandidate = parentOfTopCandidate.parentNode;
+              continue;
+            }
+            var listsContainingThisAncestor = 0;
+            for (var ancestorIndex = 0; ancestorIndex < alternativeCandidateAncestors.length; ancestorIndex++) {
+              listsContainingThisAncestor += Number(alternativeCandidateAncestors[ancestorIndex].includes(parentOfTopCandidate));
+            }
+            if (listsContainingThisAncestor >= MINIMUM_TOPCANDIDATES) {
+              topCandidate = parentOfTopCandidate;
+              break;
+            }
+            parentOfTopCandidate = parentOfTopCandidate.parentNode;
+          }
+        }
+
         // Because of our bonus system, parents of candidates might have scores
         // themselves. They get half of the node. There won't be nodes with higher
         // scores than our topCandidate, but if we see the score going *up* in the first
         // few steps up the tree, that's a decent sign that there might be more content
         // lurking in other places that we want to unify in. The sibling stuff
         // below does some of that - but only if we've looked high enough up the DOM
         // tree.
-        parentOfTopCandidate = topCandidate.parentNode;
         var lastScore = topCandidate.readability.contentScore;
         // The scores shouldn't get too low.
         var scoreThreshold = lastScore / 3;
-        while (parentOfTopCandidate && parentOfTopCandidate.readability) {
+        while (parentOfTopCandidate && parentOfTopCandidate.tagName !== "BODY") {
+          if (!parentOfTopCandidate.readability) {
+            parentOfTopCandidate = parentOfTopCandidate.parentNode;
+            continue;
+          }
           var parentScore = parentOfTopCandidate.readability.contentScore;
           if (parentScore < scoreThreshold)
             break;
```

Comment:
We don't remove the below method because these two methods are for different situations(one is the below, another one is add just added in the PR). For example, if we remove the below method the [`herald-sun-1`][herald-sun-1] test cannot be passed(some other tests cannot be passed neither) because there only one item(`<div class="story-intro">`) in `topCandidates` array is part of main content and that its score is not high enough. We need at least 3 node and theirs scores need to be high enough(75%) for the new method I just added.

```
var lastScore = topCandidate.readability.contentScore;
// The scores shouldn't get too low.
var scoreThreshold = lastScore / 3;
while (parentOfTopCandidate && parentOfTopCandidate.tagName !== "BODY") {
  if (!parentOfTopCandidate.readability) {
    parentOfTopCandidate = parentOfTopCandidate.parentNode;
    continue;
  }
  var parentScore = parentOfTopCandidate.readability.contentScore;
  if (parentScore < scoreThreshold)
    break;
  if (parentScore > lastScore) {
    // Alright! We found a better parent to use.
    topCandidate = parentOfTopCandidate;
    break;
  }
  lastScore = parentOfTopCandidate.readability.contentScore;
  parentOfTopCandidate = parentOfTopCandidate.parentNode;
}
```

[herald-sun-1]: https://github.com/mozilla/readability/blob/master/test/test-pages/herald-sun-1/source.html#L642